Validation for KubeFedCluster API

font opened the pull request:
Fixes #908.

marun commented on pkg/apis/core/v1beta1/validation/validation.go:
```diff
@@ -144,6 +149,88 @@ func ValidateFederatedTypeConfigStatus(status *v1beta1.FederatedTypeConfigStatus
+	hostname := hostURL.Hostname()
+	dnsErrs := valutil.IsDNS1123Subdomain(hostname)
+	ipErrs := valutil.IsValidIP(hostname)
+	if dnsErrs != nil && ipErrs != nil {
```
I can deduce from this logic that the hostname needs to validate either as a valid subdomain or ip, and that a validation error for both paths indicates an invalid hostname.  I'm not sure both sets of errors will make much sense to the user in the form presented, though.  The test checks for one or the other, but the user would be presented by both error conditions.  What do you think of adding a generic error message instead, similar to what `DefaultServerURL()` returns?

marun replied:
I think that's a good compromise!

font commented:
@marun thanks for your review! I've pushed an update with a new commit that addresses the feedback with the plan to squash that prior to merge (hence the hold).

> Is a ca bundle, if provided, something that could be validated?

I was originally validating the cabundle was in base64 encoding, but then discovered it is already decoded by the time the webhook gets it. So aside from that base64 encoding check, I don't think there's much else we can do.

/hold

marun commented on test/e2e/validation.go:
```diff
@@ -130,3 +132,89 @@ var _ = Describe("Core API Validation", func() {
+			client = f.Client(userAgent)
+		}
+
+		if validKFC == nil {
```
I'd vote for kfcluster over kfcls (cls reads as class to me).  I think config/conf/cfg are all readable options.

marun commented on pkg/apis/core/v1beta1/validation/validation.go:
```diff
@@ -144,6 +149,88 @@ func ValidateFederatedTypeConfigStatus(status *v1beta1.FederatedTypeConfigStatus
+	ipErrs := valutil.IsValidIP(hostname)
+	if dnsErrs != nil && ipErrs != nil {
+		combinedErrMsg := fmt.Sprintf("%s; or %s", strings.Join(ipErrs, ","), strings.Join(dnsErrs, ","))
+		allErrs = append(allErrs, field.Invalid(path, endpoint, combinedErrMsg))
```
The value used for this error message would be for the whole endpoint which could be confusing (e.g. `Invalid value: "https://example_com:8080/foo": must be a valid IP address...`). Would it be possible to use `hostname` as the value parameter instead?

marun commented on pkg/apis/core/v1beta1/validation/validation.go:
```diff
@@ -144,6 +149,88 @@ func ValidateFederatedTypeConfigStatus(status *v1beta1.FederatedTypeConfigStatus
+
+	port := hostURL.Port()
+	if port != "" {
+		portInt, _ := strconv.Atoi(port)
```
Would it make sense to report an error for a non-integer port and only check if the port is valid if the port were successfully parsed as an integer?

marun commented on pkg/apis/core/v1beta1/validation/validation.go:
```diff
@@ -144,6 +149,88 @@ func ValidateFederatedTypeConfigStatus(status *v1beta1.FederatedTypeConfigStatus
+	if port != "" {
+		portInt, _ := strconv.Atoi(port)
+		if errs := valutil.IsValidPortNum(portInt); errs != nil {
+			allErrs = append(allErrs, field.Invalid(path, endpoint, strings.Join(errs, ",")))
```
Would it be possible to provide an error message specifying that the port was in error rather than the whole endpoint message (e.g. `Invalid value: 0: port must must be between 1 and 65555)` rather than `Invalid value: "https://example_com:0/foo": must must be between 1 and 65555`)?

marun commented on pkg/apis/core/v1beta1/validation/validation_test.go:
```diff
@@ -332,6 +336,278 @@ func apiResource(apiResource *metav1.APIResource) *v1beta1.APIResource {
+		for _, endpt := range successAPIEndpoints {
+			for _, port := range successPortNums {
+				for _, path := range successURLPath {
+					if scheme == "" && path == "/path/to/endpoint" {
```
Is an empty scheme and `/` valid?

font replied:
According to `DefaultServerURL` it is. I assume it's because adding `/` still points to the root of the http server so it's interpreted the same as if it's omitted.

font commented:
@marun I've pushed a new commit to address comments. I've also updated the unit test error messages for `apiEndpoint` to be very specific, where possible, in order to verify the exact error message. PTAL.

marun commented:
@font LGTM, thank you for the cleanup!

font commented:
Squashed commits to prepare for merging.

/hold cancel

shashidharatd commented:
@font, this LGTM and thanks for all the work.
/lgtm
